ratatool-shapeless/src/main/scala/com/spotify/ratatool/shapeless/CaseClassDiffy.scala

```diff
@@ -0,0 +1,178 @@
+package com.spotify.ratatool.shapeless
+
+import com.spotify.ratatool.Command
+import com.spotify.ratatool.diffy.{BigDiffy, Delta, Diffy}
+import com.spotify.ratatool.diffy.BigDiffy.diff
+import com.spotify.scio.values.SCollection
+import shapeless._
+import shapeless.labelled.FieldType
+
+import scala.reflect.ClassTag
+
+sealed trait MapEncoder[A] {
+  def toMap(in: A): Map[String, Any]
+}
+
+object MapEncoder {
+  def apply[A](implicit encoder: MapEncoder[A]): MapEncoder[A] =
+    encoder
+
+  def createEncoder[A](func: A => Map[String, Any]): MapEncoder[A] = new MapEncoder[A] {
+    override def toMap(in: A): Map[String, Any] = func(in)
+  }
+
+  implicit def stringEncoder[K <: Symbol](implicit witness: Witness.Aux[K]):
+  MapEncoder[FieldType[K, String]] = {
+    val name = witness.value.name
+    createEncoder { v =>
+      Map(name -> v.self)
+    }
+  }
+
+  implicit def intEncoder[K <: Symbol](implicit witness: Witness.Aux[K]):
+  MapEncoder[FieldType[K, Int]] = {
+    val name = witness.value.name
+    createEncoder { v =>
+      Map(name -> v.self)
+    }
+  }
+
+  implicit def longEncoder[K <: Symbol](implicit witness: Witness.Aux[K]):
+  MapEncoder[FieldType[K, Long]] = {
+    val name = witness.value.name
+    createEncoder { v =>
+      Map(name -> v.self)
+    }
+  }
+
+  implicit def booleanEncoder[K <: Symbol](implicit witness: Witness.Aux[K]):
+  MapEncoder[FieldType[K, Boolean]] = {
+    val name = witness.value.name
+    createEncoder { v =>
+      Map(name -> v.self)
+    }
+  }
+
+  implicit def floatEncoder[K <: Symbol](implicit witness: Witness.Aux[K]):
+  MapEncoder[FieldType[K, Float]] = {
+    val name = witness.value.name
+    createEncoder { v =>
+      Map(name -> v.self)
+    }
+  }
+
+  implicit def doubleEncoder[K <: Symbol](implicit witness: Witness.Aux[K]):
+  MapEncoder[FieldType[K, Double]] = {
+    val name = witness.value.name
+    createEncoder { v =>
+      Map(name -> v.self)
+    }
+  }
+
+  implicit val hnilEncoder: MapEncoder[HNil] =
+    createEncoder[HNil](n => Map(HNil.toString -> HNil))
+
+  implicit def seqEncoder[K <: Symbol, V](implicit witness: Witness.Aux[K]):
+  MapEncoder[FieldType[K, Seq[V]]] = {
+    val name = witness.value.name
+    createEncoder { v =>
+      Map(name -> v.toIndexedSeq)
+    }
+  }
+
+  implicit def listEncoder[K <: Symbol, V](implicit witness: Witness.Aux[K]):
+  MapEncoder[FieldType[K, List[V]]] = {
+    val name = witness.value.name
+    createEncoder { v =>
+      Map(name -> v.toIndexedSeq)
+    }
+  }
+
+  /*
+  * null.asInstanceOf[V] derives the default value of primitive types
+  * null.asInstanceOf[Int] = 0
+  * null.asInstanceOf[Boolean] = false
+  * null.asInstanceOf[String] = null
+  * */
+  implicit def optionEncoder[K <: Symbol, V](implicit witness: Witness.Aux[K]):
+  MapEncoder[FieldType[K, Option[V]]] = {
+    val name = witness.value.name
+    createEncoder { v =>
+      Map(name -> v.getOrElse(null.asInstanceOf[V]))
+    }
+  }
+
+  //scalastyle:off
```

Comment:
Missed this, should this be left in here? Also the following line looks a bit too long

Reply:
I was thinking between disable line-length style checking and using type alias to shorten the code :

```
type ME[T] = MapEncoder[T]
  type FT[K, V] = FieldType[K, V]
  type Aux[H, R] = LabelledGeneric.Aux[H, R]

  implicit def hlistEncoder0[K <: Symbol, H, T <: HList](implicit hEnc: Lazy[ME[FT[K, H]]],
                                                         tEnc: Lazy[ME[T]]): ME[FT[K, H] :: T] = {
    createEncoder[FieldType[K, H] :: T]{ in =>
      hEnc.value.toMap(in.head) ++ tEnc.value.toMap(in.tail)
    }
  }

  implicit def hListEncoder1[K <: Symbol, H, T <: HList, R <: HList](implicit
                                                                      wit: Witness.Aux[K],
                                                                      gen: Aux[H, R],
                                                                      encoderH: Lazy[ME[R]],
                                                                      encoderT: Lazy[ME[T]]
                                                                     ): ME[FT[K, H] :: T] =
    createEncoder(in =>
      encoderT.value.toMap(in.tail) ++ Map(wit.value.name -> encoderH.value.toMap(gen.to(in.head))))
```